simulate/sam.go

```diff
@@ -0,0 +1,143 @@
+package simulate
+
+import (
+	"fmt"
+	"github.com/vertgenlab/gonomics/cigar"
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/numbers"
+	"github.com/vertgenlab/gonomics/sam"
+	"math/rand"
+	"strings"
+)
+
+// IlluminaSam generates pseudorandom paired reads evenly distributed along the input ref sequence using typical illumina parameters.
+func IlluminaSam(refName string, ref []dna.Base, numPairs int) []sam.Sam {
+	return simulatePairedSam(refName, ref, numPairs, 150, 50, 50)
+}
+
+// simulatePairedSam generates a pair of sam reads randomly distributed across the input ref.
+func simulatePairedSam(refName string, ref []dna.Base, numPairs, readLen, avgInsertSize int, avgInsertSizeStdDev float64) []sam.Sam {
+	reads := make([]sam.Sam, numPairs*2)
+
+	var insertSize, midpoint, startFor, startRev, endFor, endRev int
+	for i := 0; i < len(reads); i += 2 {
+		insertSize = int(numbers.SampleInverseNormal(float64(avgInsertSize), avgInsertSizeStdDev))
+		midpoint = numbers.RandIntInRange(-2*readLen, len(ref)+(2*readLen))
```

Comment:
I am having trouble following the logic here.  I would have thought that the midpoint would have to come in from the ends (to allow space for the sides of the fragment, without going off the end of the chrom), but it seems like here we are allowing the midpoint to be beyond both ends of the chromosome.

Reply:
This was some gymnastics I did to get even coverage across the contig. My initial code generated a gradient where coverage increases at the start of the contig and decreases slowly towards the end of the contig. This lets a read start past the ends of the contig (and get trimmed if it goes off the contig) to ensure even coverage.

Reply:
Hmm I think it might depend on your fragmentation method. With something like motif-based enzymatic fragmentation you would probably get major depletion of the fragment ends, but with sonication-based fragmentation you may get a bit of depletion but I would guess the ends would be mostly intact. Either way, I've changed the code to take the midpoint from [0, len(ref)) but left in a commented line for the old method in case it becomes relevant.